iree/vm/list.c

```diff
@@ -686,3 +686,124 @@ iree_status_t iree_vm_list_register_types(void) {
   iree_vm_list_descriptor.type_name = iree_make_cstring_view("vm.list");
   return iree_vm_ref_register_type(&iree_vm_list_descriptor);
 }
+
+#define std_swap(a, b) \
+  do {                 \
+    typeof(a) t = a;   \
+    a = b;             \
+    b = t;             \
+  } while (0);
+
+static bool iree_vm_list_is_the_same_type(const iree_vm_list_t* list_a,
+                                          const iree_vm_list_t* list_b) {
+  return list_a->element_type.value_type == list_b->element_type.value_type &&
+         list_a->element_type.ref_type == list_b->element_type.ref_type;
+}
+
+static bool iree_vm_list_is_compatiable_allocator(
+    const iree_vm_list_t* list_a, const iree_vm_list_t* list_b) {
+  // Reject if both lists are allocated on the stack. The allocator is
+  // iree_allocator_null if the list is allocated on the stack.
+  //
+  // benvanik: swapping between stack allocated lists is probably fine, though
+  // to be conservative we could just fail.
+  if (iree_allocator_is_null(list_a->allocator) ||
+      iree_allocator_is_null(list_b->allocator))
+    return false;
+
+  return list_a->allocator.ctl == list_b->allocator.ctl;
+}
+
+iree_status_t iree_vm_list_swap(iree_vm_list_t* list_a,
+                                iree_vm_list_t* list_b) {
+  if (!iree_vm_list_is_the_same_type(list_a, list_b))
+    return iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                            "try to swap two different types of lists.");
+
+  if (!iree_vm_list_is_compatiable_allocator(list_a, list_b))
+    return iree_make_status(
+        IREE_STATUS_FAILED_PRECONDITION,
+        "try to swap two incompatible allocators of lists.");
+
+  // FIXME: can we exchange ref_object?
+  std_swap(list_a->ref_object, list_b->ref_object);
+  // FIXME: can we exchange allocator.self?
+  std_swap(list_a->allocator.self, list_b->allocator.self);
```

Comment:
we can't exchange ref_objects (and wouldn't want to, as the intent here is that the lists stay the same and only their storage changes) - and we don't need to change the allocator (as we ensure the allocators are the same) - so fine to remove these lines